udp.go

```diff
@@ -9,6 +9,34 @@ import (
 	"golang.org/x/net/ipv6"
 )
 
+type SessionUDPFactory interface {
```

Comment:
I'm pretty certain this interface is entirely unnecessary. Using the `Reader` interface, it's possible to provide an entirely custom `ReadUDP` implementation, which is the only place that `SessionUDP` structs are returned.

Reply:
I started with this assumption, but soon found out that `ReadUDP` implemenations have no access to `SessionUDP` structs (nothing there is exported).

Reply:
Furthermore, even if it was possible to return a pointer to a struct `SessionUDP` from a new ReadUDP() implementation, the rest of the code would need to know what to do with the UDP control data therein (e.g., IP_ORIGDSTADDR needed for TPROXY), or at least pass it to my Writer implemenation.

Reply:
Oh okay, I think I understand your problem here. I think a much smaller PR will make your desired use-case possible, even if not the cleanest. Am I right in thinking your main problem is not being able to pass extra information from the reader code to the writer code?